runhouse/rns/rns_client.py

```diff
@@ -664,20 +665,38 @@ def _get_logs(self):
         cleaned_log_lines = [ColoredFormatter.format_log(line) for line in log_lines]
         return " ".join(cleaned_log_lines)
 
-    async def send_cluster_logs_to_den(self, cluster_uri: str, api_server_url: str):
+    def _generate_logs_file_name(self):
+        current_timestamp = datetime.datetime.utcnow().strftime("%Y_%m_%d_%H_%M_%S")
+        return f"{current_timestamp}_{S3_LOGS_FILE_NAME}"
+
+    async def send_cluster_logs_to_den(
+        self, cluster_uri: str, api_server_url: str, prev_end_log_line: int
+    ):
 
         try:
-            latest_logs = self._get_logs()
-            logs_data = {"file_name": S3_LOGS_FILE_NAME, "logs": latest_logs}
+
+            # setting to a list, so it will be easier to get the end line num  + the logs delta to send to den.
+            latest_logs = self._get_logs().split("\n")
+
+            new_end_log_line = len(latest_logs) - prev_end_log_line - 1
+            # minus 1 because we start counting logs from 0.
+
+            logs_to_den = " ".join(latest_logs[prev_end_log_line:])
+            logs_data = {
+                "file_name": self._generate_logs_file_name(),
+                "logs": logs_to_den,
+                "start_line": prev_end_log_line,
+                "end_line": new_end_log_line,
+            }
 
             post_logs_resp = requests.post(
                 f"{api_server_url}/resource/{cluster_uri}/logs",
                 data=json.dumps(logs_data),
                 headers=self.request_headers(),
             )
-            return post_logs_resp.status_code
+            return post_logs_resp.status_code, prev_end_log_line, new_end_log_line
         except Exception as e:
             logger.error(
```

Comment:
if the request to den fails (not 200), this exception won't actually be thrown right? this is meant to catch the other logic leading up to the request?

Reply:
correct :)

Reply:
So small bit but the log message in the exception handler maybe can be more accurate, bc it's not a problem sending the logs to done it's a problem generating the logs before they are sent